### PR TITLE
System layer implementation.

### PR DESCRIPTION
**Pull request by Kerilk**
This is a tentative implementation of system wide layers. Those layers are supposed to always be loaded before user-specified layers, meaning they will be deeper in the layer stack.

On Linux/MacOs/Android:
 - By default layers are configured in `/etc/OpenCL/layers/` (`/system/vendor/Khronos/OpenCL/layers` on Android);
 - This path can be overiden using the `OPENCL_LAYER_PATH` environment variable (similar to `OCL_ICD_VENDORS`);
 - Similarly to ICDs, layers are configured in a text file ending in `.lay`;
 - As ordering of layers is important, they are alphanumerically sorted by filename, meaning that prefixing `.lay` files with priorities should allow the desired ordering (akin to many configuration systems).

On Windows:
 - Layers are configured through the registry key: `SOFTWARE\Khronos\OpenCL\Layers`;
 - Similarly to ICDs, the layers are added through a `REG_DWORD` values;
 - Contrary to ICDs, the value of the keys can be non zero and will be used as a priority value to sort them.

Many of these choices are arbitrary and can be discussed, so feel free to rip through this proposal.

I tried to respect each system coding practices but may have missed the mark. Especially, for the windows code path, I didn't feel confident enough to refactor the two variable list implementations and come up with something satisfactory. I took more liberties refactoring the *nix code path.

I extensively tested the Linux implementation by generating different print layers with slightly modified output and trying different ordering and combination of system and user specified layers (using the `OPENCL_LAYERS` environment variable), but the diff is quite big so I may have let slip something through. I could only validate the windows code path through building it using GitHub actions, so it would be good that it is given a thorough look/testing.

PS: I am unsure I am correctly understanding the Windows safe mode issue: as far as I understand it prohibits the usage of environment variables, preventing any user configuration of the loader through this mean. Is that correct? If it is it would prevent any device filtering using environment variables unless some variables are treated differently.

### CODE REVIEW

File: loader/linux/icd_linux.c
Comment (jlewis-austin) [inline loader/linux/icd_linux.c:213]: Is it important to to distinguish these as system layers? I'm thinking that a user can choose to use the file-based mechanism for their own layers if they want to, e.g. if using environment variables is inconvenient on Android.

File: README.md
Comment (jlewis-austin) [inline README.md:146]: OCL_LAYERS vs OPENCL_LAYERS are going to be confusing to remember. Maybe OCL_LAYER_LIBS and OCL_LAYER_DIR (assuming it's not too late to change the existing name)?

**Comment by Kerilk** [comment]
@jlewis-austin Thanks for the feedback, it is greatly appreciated.

File: loader/linux/icd_linux.c
Comment (Kerilk) [reply loader/linux/icd_linux.c:213]: Indeed the user can use them, as long as they can install the libraries in the default folder. On a classic linux distribution that would require root access, unless the user overrides the default location through environment variables. This naming was sparked from a discussion about layers that could be configured in safe mode as well on windows platform, so with a system wide scope and independent from users. We can discuss giving them a new name.

File: README.md
Comment (Kerilk) [reply README.md:146]: Good catch on the copy paste, I'll fix that.
For the naming of the environment variables, you are correct that it is going to be difficult to converge on something satisfactory. The original name of OPENCL_LAYERS was OCL_ICD_LAYERS, but was changed by request of the working group. Unfortunately, rolling it back/changing it is not possible anymore. We can still change OCL_LAYERS to OCL_LAYER_LIBS or OPENCL_LAYER_LIBS.

**Comment by Kerilk** [comment]
Non trivial rebase on top of master. Should be tested thoroughly.

File: loader/linux/icd_linux.c
Comment (MathiasMagnus) [inline loader/linux/icd_linux.c:48]: This comment became out-of-sync.
```suggestion
    // make sure the file name ends in `extension` (eg. .icd, or .lay)
```

File: loader/windows/icd_windows.c
Comment (MathiasMagnus) [inline loader/windows/icd_windows.c:136]: If we print out-of-memory errors on Linux (in which case printing to the console is problematic on it's own), we should strive to be consistent and do the same on Windows.
```suggestion
        if (!pNewBegin)
        {
            KHR_ICD_TRACE("Failed allocate space for Layers array\n");
            result = FALSE;
        }
```

File: loader/windows/icd_windows.c
Comment (MathiasMagnus) [inline loader/windows/icd_windows.c:154]:
```suggestion
        if (!pWinLayerEnd->szName)
        {
            KHR_ICD_TRACE("Failed allocate space for Layer file path\n");
            result = FALSE;
        }
```

**Review by MathiasMagnus — CHANGES_REQUESTED** [review]
Overall LGTM, minor changes noted, but overall it just needs to be rebased on latest master (which since received the SDK base changes, but should apply cleanly, GitHub may even do it on it's own).

**Comment by Kerilk** [comment]
> Overall LGTM, minor changes noted, but overall it just needs to be rebased on latest master (which since received the SDK base changes, but should apply cleanly, GitHub may even do it on it's own).

Thanks for the review. I've rebased the branch on `master`, so if you want to use it in the Layers CI you can do so, and it should be easy to switch back to the `master` branch once those changes have been merged. Especially, testing the layer sorting algorithm for windows could be useful. I did test the lexicographic sorting in Linux though, but Window's version works using the value of the layers in the registry.

**Comment by Kerilk** [comment]
Re-based on main.

**Review by MathiasMagnus — APPROVED** [review]

File: README.md
Comment (Kerilk) [reply README.md]: I implemented the change @bashbaug suggested, as it seems a good choice.

File: loader/linux/icd_linux.c
Comment (bashbaug) [reply loader/linux/icd_linux.c:220]: Can we check what Vulkan is doing to see if we are consistent?

**Comment by Kerilk** [comment]
I know I said otherwise at the working group meeting, but thoroughly testing the feature (variables impact, load ordering) would require the cllayerinfo tool which depends on this PR, but still needs some work. I propose to merge this one once the issues above have been resolved, and will implement the CI in  https://github.com/KhronosGroup/OpenCL-ICD-Loader/pull/175
@MathiasMagnus @bashbaug does that seem a reasonable path forward?

**Comment by bashbaug** [comment]
Discussed in the September 5th teleconference.  Agreed to these changes in principle, and we're aiming to merge in the next couple of days.

TODOs:
- Independently test on Linux.
- Detailed code review - ideally we get one or two more approvals before merging.

**Comment by Kerilk** [comment]
@MathiasMagnus , @bashbaug the rebase is done, and the CI passes. I retested on my Ubuntu (22.04) and both `OPENCL_LAYER_PATH` and the default path (`/etc/OpenCL/layers/`) behaved as expected.

File: loader/linux/icd_linux.c
Comment (bashbaug) [inline loader/linux/icd_linux.c:131]: I don't have a strong opinion about this but I do want to confirm that strcoll is intended here rather than strcmp, just because no other part of the ICD loader uses strcoll.  If it is intended, I'd suggest adding a comment to avoid confusion in the future.

**Review by bashbaug — COMMENTED** [review]
This looks really great!  I tested layers both in the default `/etc/OpenCL/layers` and `OPENCL_LAYER_PATH` environment variable, including layers in different orders, and things worked as expected.

Would you mind bumping `OPENCL_ICD_LOADER_VERSION_REV` after this change?

I have one other question but it is not critical.

**Comment by bashbaug** [comment]
Looks like this needs a rebase now also - hopefully not too bad.

**Comment by MathiasMagnus** [comment]
[Windows](https://github.com/StreamHPC/OpenCL-Layers/actions/runs/3027003399) and [Linux](https://github.com/StreamHPC/OpenCL-Layers/actions/runs/3027003405) CI running.

File: loader/linux/icd_linux.c
Comment (Kerilk) [reply loader/linux/icd_linux.c:131]: Yes it is intended, this is the way directory entries are sorted alphanumericaly on unixes (as far as I understand such things). I wanted our sorting to behave with minimum surprise for the user. I will add a comment.

**Comment by Kerilk** [comment]
> This looks really great! I tested layers both in the default `/etc/OpenCL/layers` and `OPENCL_LAYER_PATH` environment variable, including layers in different orders, and things worked as expected.
> 
> Would you mind bumping `OPENCL_ICD_LOADER_VERSION_REV` after this change?

Will do, and I will rebase.

File: loader/linux/icd_linux.c
Comment (Kerilk) [reply loader/linux/icd_linux.c:131]: Digging deeper, maybe I should have used libc alphasort directly, didn't know about it...

Comment (bashbaug) [reply loader/linux/icd_linux.c:131]: I see, thanks!  Adding a comment is fine for now - no need to make major changes.

**Comment by Kerilk** [comment]
@bashbaug @Kerilk Thanks again for the reviews, I added the reasoning about `strcoll` usage, rebased on the latest main, and incremented windows version number.

**Pull request merged**
